File: src/password/breach_checker.py
"""
Password breach checker.
Checks if passwords have been exposed in data breaches.
Uses a zero-knowledge approach to maintain privacy.
"""

import hashlib
import requests
import sys
import os

# Add the project root to the Python path
sys.path.append(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))

from config.config import HIBP_API_URL, HIBP_USER_AGENT


class BreachChecker:
    """Checks passwords against known data breaches securely"""
    
    def __init__(self):
        """Initialize breach checker"""
        self.api_url = HIBP_API_URL
        self.user_agent = HIBP_USER_AGENT
        self.cache = {}  # Simple in-memory cache
    
    def check_password(self, password, use_cache=True):
        """
        Check if a password has been exposed in known data breaches
        Uses the k-anonymity model from Have I Been Pwned
        
        Args:
            password (str): Password to check
            use_cache (bool): Whether to use the cache
            
        Returns:
            dict: Result of the check
        """
        try:
            # Convert empty passwords to a string to avoid errors
            if not password:
                return {
                    'compromised': False,
                    'count': 0,
                    'message': 'Empty password not checked'
                }
            
            # Create a SHA-1 hash of the password
            password_hash = hashlib.sha1(password.encode('utf-8')).hexdigest().upper()
            
            # Check cache first if enabled
            if use_cache and password_hash in self.cache:
                return self.cache[password_hash]
            
            # Use the first 5 characters as the prefix
            prefix = password_hash[:5]
            suffix = password_hash[5:]
            
            # Make API request
            response = self._make_api_request(prefix)
            
            # Parse response
            result = self._parse_response(response, suffix)
            
            # Cache result
            if use_cache:
                self.cache[password_hash] = result
            
            return result
        except Exception as e:
            # Fail gracefully - always assume password is safe if service fails
            return {
                'compromised': False,
                'count': 0,
                'message': f'Error checking password: {str(e)}'
            }
    
    def generate_breach_notification(self, result):
        """
        Generate a human-readable notification about the breach check result
        
        Args:
            result (dict): Result from check_password
            
        Returns:
            str: Human-readable notification
        """
        if 'status' in result and result['status'] == 'error':
            return f"Unable to check breach status: {result.get('message', 'Unknown error')}"
        
        if result.get('compromised'):
            count = result.get('count', 0)
            if count == 1:
                return "⚠️ WARNING: This password has been exposed in a data breach!"
            else:
                return f"⚠️ WARNING: This password has been found in {count} data breaches!"
        else:
            return "✅ Not found in known data breaches."
    
    def _make_api_request(self, prefix):
        """
        Make an API request to the breach checking service
        
        Args:
            prefix (str): The hash prefix to check
            
        Returns:
            str: The API response
        """
        headers = {
            'User-Agent': self.user_agent
        }
        
        url = f"{self.api_url}/{prefix}"
        
        try:
            response = requests.get(url, headers=headers, timeout=5)
            response.raise_for_status()
            return response.text
        except requests.exceptions.RequestException as e:
            raise Exception(f"API request failed: {str(e)}")
    
    def _parse_response(self, response, suffix):
        """
        Parse the API response to check if the password suffix is present
        
        Args:
            response (str): The API response
            suffix (str): The hash suffix to look for
            
        Returns:
            dict: The result of the check
        """
        lines = response.splitlines()
        
        for line in lines:
            parts = line.split(':')
            if len(parts) == 2:
                found_suffix = parts[0].strip()
                count = int(parts[1].strip())
                
                if found_suffix == suffix:
                    return {
                        'compromised': True,
                        'count': count,
                        'message': f'This password has been exposed in {count} data breaches'
                    }
        
        return {
            'compromised': False,
            'count': 0,
            'message': 'This password was not found in known data breaches'
        }


# Singleton instance
breach_checker = BreachChecker() 

File: src/core/password/breachChecker.js
const crypto = require('crypto');
const https = require('https');

/**
 * Service for checking if passwords have been compromised in known data breaches
 * using the "Have I Been Pwned" API with k-anonymity model for privacy
 */
class BreachChecker {
  constructor() {
    this.pwnedApiUrl = 'https://api.pwnedpasswords.com/range/';
    this.userAgent = 'SecurePasswordManager/1.0';
  }

  /**
   * Check if a password has been compromised in known data breaches
   * @param {string} password - The password to check
   * @returns {Promise<Object>} - Breach check results
   */
  async checkPassword(password) {
    try {
      // Generate SHA-1 hash of the password
      const sha1Hash = this.hashPassword(password);
      
      // Split the hash into prefix and suffix for k-anonymity
      const prefix = sha1Hash.substring(0, 5);
      const suffix = sha1Hash.substring(5).toUpperCase();
      
      // Get all hash suffixes that match the prefix
      const matchingHashes = await this.getMatchingHashes(prefix);
      
      // Check if our suffix is in the results
      const match = matchingHashes.find(hash => hash.suffix === suffix);
      
      return {
        breached: match !== undefined,
        occurrenceCount: match ? match.count : 0,
        message: this.generateMessage(match)
      };
    } catch (error) {
      console.error('Error checking for breaches:', error);
      return {
        breached: false,
        occurrenceCount: 0,
        message: 'Unable to check password against known breaches',
        error: error.message
      };
    }
  }

  /**
   * Hash a password using SHA-1 (for HIBP API compatibility only)
   * @param {string} password - The password to hash
   * @returns {string} - SHA-1 hash of the password
   */
  hashPassword(password) {
    return crypto
      .createHash('sha1')
      .update(password)
      .digest('hex')
      .toUpperCase();
  }

  /**
   * Query HIBP API with k-anonymity to get matching hash suffixes
   * @param {string} hashPrefix - First 5 characters of password hash
   * @returns {Promise<Array>} - Matching hash suffixes with counts
   */
  async getMatchingHashes(hashPrefix) {
    return new Promise((resolve, reject) => {
      const requestOptions = {
        headers: {
          'User-Agent': this.userAgent
        }
      };
      
      const req = https.get(
        `${this.pwnedApiUrl}${hashPrefix}`,
        requestOptions,
        (res) => {
          if (res.statusCode !== 200) {
            reject(new Error(`HIBP API request failed: ${res.statusCode}`));
            return;
          }
          
          let data = '';
          res.on('data', (chunk) => {
            data += chunk;
          });
          
          res.on('end', () => {
            try {
              // Parse response into hash suffix and count pairs
              const results = data
                .split('\r\n')
                .filter(line => line.length > 0)
                .map(line => {
                  const [suffix, countStr] = line.split(':');
                  return {
                    suffix,
                    count: parseInt(countStr, 10)
                  };
                });
              
              resolve(results);
            } catch (error) {
              reject(error);
            }
          });
        }
      );
      
      req.on('error', (error) => {
        reject(error);
      });
      
      req.end();
    });
  }

  /**
   * Generate a message based on breach check results
   * @param {Object|undefined} match - Matching hash data
   * @returns {string} - Result message
   */
  generateMessage(match) {
    if (!match) {
      return 'Good news! This password hasn\'t been found in any known data breaches.';
    }
    
    if (match.count === 1) {
      return 'Warning: This password has been found in 1 data breach. It is recommended to choose a different password.';
    }
    
    let severityLevel = 'Warning';
    if (match.count > 1000) {
      severityLevel = 'Critical';
    } else if (match.count > 100) {
      severityLevel = 'Severe';
    } else if (match.count > 10) {
      severityLevel = 'High Risk';
    }
    
    return `${severityLevel}: This password has been found in ${match.count.toLocaleString()} data breaches. You should definitely choose a different password.`;
  }
}

module.exports = new BreachChecker(); 